hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/DotNode.java

```diff
@@ -394,11 +394,22 @@ private void dereferenceEntity(
 			// thus we need to generate a join unless the association is non-nullable and
 			// parent refers to the associated entity's PK (because 'our' table would know the FK).
 			parentAsDotNode = (DotNode) parent;
+			// our parent is another dot node, meaning we are being further de-referenced.
+			// depending on the exact de-reference we may need to generate a physical join.
+
 			property = parentAsDotNode.propertyName;
-			joinIsNeeded = generateJoin && (
-					entityType.isNullable() ||
-							!isReferenceToPrimaryKey( parentAsDotNode.propertyName, entityType )
-			);
+
+			if ( generateJoin ) {
+				if ( implicitJoin && ( toOneType.hasNotFoundAction() || toOneType.isNullable() ) ) {
+					joinIsNeeded = true;
+				}
+				else {
+					joinIsNeeded = !isPropertyEmbeddedInJoinProperties( parentAsDotNode.propertyName );
+				}
+			}
```

Comment:
Note to self, I think it might best to use this here instead.

```suggestion
			if ( generateJoin ) {
				if ( implicitJoin && toOneType.isNullable() ) {
					joinIsNeeded = true;
				}
				else {
					joinIsNeeded = !isReferenceToPrimaryKey( parentAsDotNode.propertyName );
				}
			}
```